server/src/main/java/org/opensearch/search/optimization/filterrewrite/DateHistogramAggregatorBridge.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.optimization.filterrewrite;
+
+import org.apache.lucene.document.LongPoint;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.opensearch.common.Rounding;
+import org.opensearch.index.mapper.DateFieldMapper;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.search.aggregations.bucket.histogram.LongBounds;
+import org.opensearch.search.aggregations.support.ValuesSourceConfig;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.OptionalLong;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import static org.opensearch.search.optimization.filterrewrite.TreeTraversal.multiRangesTraverse;
+
+/**
+ * For date histogram aggregation
+ */
+public abstract class DateHistogramAggregatorBridge extends AggregatorBridge {
+
+    protected boolean canOptimize(ValuesSourceConfig config) {
+        if (config.script() == null && config.missing() == null) {
+            MappedFieldType fieldType = config.fieldType();
+            if (fieldType instanceof DateFieldMapper.DateFieldType) {
+                if (fieldType.isSearchable()) {
+                    this.fieldType = fieldType;
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    protected void buildRanges(SearchContext context) throws IOException {
+        long[] bounds = Helper.getDateHistoAggBounds(context, fieldType.name());
+        optimizationContext.setRanges(buildRanges(bounds));
+    }
+
+    @Override
+    public void prepareFromSegment(LeafReaderContext leaf) throws IOException {
+        long[] bounds = Helper.getSegmentBounds(leaf, fieldType.name());
+        optimizationContext.setRangesFromSegment(buildRanges(bounds));
```

Comment:
return the ranges here for the context to fetch instead of setting state on the context.

Reply:
Updated as this method actually should be pkg-private